src/ExecBackends/ExecSpace.hpp

```diff
@@ -0,0 +1,450 @@
+  {
+    return MemBackendHost("HOST");
+  };
+
+private:
+  std::string mem_space_;
+};
+#endif //HIOP_USE_RAJA //HIOP_USE_UMPIRE
+
+#ifdef HIOP_USE_CUDA
+/// Cuda memory backend for device memory space that is implemented using Cuda API
+struct MemBackendCuda
+{
+  /**
+   * Constructor taking a memory space as input; provided for exchangeability with
+   * other memory backends.
+   */
+  MemBackendCuda(std::string mem_space = "DEVICE")
+  {
+    assert(mem_space == "DEVICE");
+  }
+  
+  /// For now does not support host memory space (but can/will be implemented).
+  inline static bool is_host() { return false; }
+
+  inline static bool is_device() { return true; }
+
+  using MemBackendHost = MemBackendCpp;
+  /// Returns a backend set up for host memory space
+  inline static MemBackendHost new_backend_host()
+  {
+    return MemBackendHost();
+  };
+};
+#endif //HIOP_USE_CUDA
+
+#ifdef HIOP_USE_HIP
+/// Cuda memory backend for device memory space that is implemented using Hip API
+struct MemBackendHip
+{
+  /**
+   * Constructor taking a memory space as input; provided for exchangeability with 
+   * other memory backends.
+   */
+  MemBackendHip(std::string mem_space = "DEVICE")
+  {
+    assert(mem_space == "DEVICE");
+  }
+
+  /// For now does not support host memory space (but can/will be implemented).
+  inline static bool is_host() { return false; }
+
+  inline static bool is_device() { return true; }
+
+  using MemBackendHost = MemBackendCpp;
+  /// Returns a backend set up for host memory space
+  inline static MemBackendHost new_backend_host()
+  {
+    return MemBackendHost();
+  };
+};
+#endif //HIOP_USE_HIP
+
+///////////////////////////////////////////////////////////////////////////////////////
+// Execution policies
+///////////////////////////////////////////////////////////////////////////////////////
+
+/// Standard C++ sequential execution
+struct ExecPolicySeq
+{
+};
+
+#ifdef HIOP_USE_CUDA
+struct ExecPolicyCuda
+{
+  ExecPolicyCuda()
+    : bl_sz_binary_search(16),
+      bl_sz_vector_loop(256)
+  {
+  }
+  /** Block size for kernels performing binary search (e.g., updating or getting diagonal 
+   *  in CSR CUDA matrices. Default value 16.
+   */
+  unsigned short int bl_sz_binary_search;
+  /// Block size for kernels performing element-wise ops on vectors. Default 256.
+  unsigned short int bl_sz_vector_loop;
+};
+#endif
+
+#ifdef HIOP_USE_HIP
+struct ExecPolicyHip
+{
+  unsigned short int num_blocks_vector;
+  unsigned short int num_blocks_search;
+};
+#endif
+
+///////////////////////////
+// RAJA execution policies
+//////////////////////////
+#ifdef HIOP_USE_RAJA
+
+#ifdef HIOP_USE_CUDA
+struct ExecPolicyRajaCuda
+{
+  //empty since no runtime info is stored
+};
+#endif
+
+#ifdef HIOP_USE_HIP
+struct ExecPolicyRajaHip
+{
+  //empty since no runtime info is stored
+};
+#endif
+
+//RAJA OMP execution policies backend
+#if !defined(HIOP_USE_CUDA) && !defined(HIOP_USE_HIP)
```

Comment:
```suggestion
#ifdef HIOP_USE_CUDA
struct ExecPolicyRajaCuda
{
  //empty since no runtime info is stored
};
#elifdef HIOP_USE_HIP
struct ExecPolicyRajaHip
{
  //empty since no runtime info is stored
};
#else
```
`#elifdef` is C++23, so please don't apply this, but I was curious if the syntax could be cleaner.

Reply:
currently cuda and hip is exclusive, but not necessarily the case in the future